### DISCUSSION TITLE
How to fit nonlinear curves using inlabru with gaussian or non-gaussian response data

### DISCUSSION THREAD

**Discussion by ribailey** [inlabru-org/inlabru#146] [Q&A]
Hi,

I'm starting the process of fitting various nonlinear curves to data using inlabru, and would like some help in particular for setting up 'lik' and 'bru' for binary response data, but also generalized to other non-gaussian responses. Below is code for generating binary data, and also continuous data with gaussian residuals, along a 1D transect using a function for a normal distribution (one of the curves I want to fit), plus a basic fitting of the gaussian data using default priors (I haven't yet attempted to specify non-default priors for any model).

First, creating the dataset:

```
library(data.table)

#1000 sample sites along a linear transect, with 10 sampled individuals per site.
Nsites=1000
Nindpersite=10

meand=0                                   #Mean of the curve
sdd=1                                        #sd of the curve
h=1                                            #height of the peak
x=seq(-10,10,length=Nsites)     #Position of sample sites along transect
varsamp=0.01                            #error variance for gaussian response data
precision=1/varsamp
sdsamp=sqrt(varsamp)

inputdt=data.table(expand.grid(x=x,ind=seq(1:Nindpersite)))      #Creating a data.table with one row per sample
inputdt[,indref:=paste("I",x,ind,sep="_")]                                       #Individual references (not used here)
inputdt[,h:=h]
inputdt[,meand:=meand]
inputdt[,sdd:=sdd]
inputdt[,log_sdd:=log(sdd)]  #Based on a previous answer from Finn, exp(log_sdd) is used rather than sdd, and log_sdd is
                            #estimated: "...the parameterisation doesn't have a unique solution for sdd; the predictor isn't
                            #monotonic in sdd. But this is easily fixed here, since you can use a log-transformation for the 
                            #sdd parameterisation, with exp(log_sdd) instead of sdd."

#creating continuous response data with gaussian residuals.
inputdt[,normcurve:=h*exp(-((x - meand)^2/(2*exp(log_sdd)^2))) + rnorm(.N,0,sdsamp)]

#Now making some binary data using the same function.
inputdt[,binarydat:=rbinom(.N,1,prob=
 h*exp(-((x - meand)^2/(2*exp(log_sdd)^2)))
                          )]

#The blue line in the plot below is a smooth.spline fitted through the binary data (binary data not plotted).

plot(normcurve~x,data=inputdt,pch=16,cex=0.3,col="grey")    #The gaussian data#
points((h*exp(-((x - meand)^2/(2*exp(log_sdd)^2))))~x,data=inputdt,col="red",pch=16,cex=0.3)   #The 'true' curve#
lines(smooth.spline(inputdt$binarydat~inputdt$x),col="blue")   #spline fitted to binary data#
```

Now fitting the continuous data using default priors.

```
lik <- inlabru::like(
  family = "gaussian", data = inputdt,
  formula = normcurve~h*exp(-((x - meand)^2/(2*exp(log_sdd)^2)))
)

fita <- bru(~ h(1) + meand(1) + log_sdd(1) - 1, lik,
            options = list(bru_initial = list(h = 2)))                  #I actually ran it successfully without setting bru_initial; h can potentially be 0 or negative#
```

My first objective is to successfully fit the binary data created above. Example code would be great, as my knowledge of INLA/inlabru is very patchy.

More generally, I hope to learn how to specify the priors etc. for the gaussian, binary, and other data distributions for this kind of nonlinear model fitting.

Here are initial responses from Finn posted on r-inla today (many thanks!): 

"For prior specifications, it works as in plain INLA, which is that components with fixed effects take arguments mean.linear and prec.linear, and for hyperparameters for structured random effect components, hyper = list(...). In your example, you only have "fixed effect" type parameters, so use mean.linear and prec.linear to specify their N(m,prec^(-1) ) priors.

For non-gaussian priors for those, you need to build that into the transformations (e.g. you now have exp(log_sdd) for the [std.dev](http://std.dev/). parameter, giving a log-normal prior on that. For an exponential prior, you would need the effect of qexp(pnorm(internal_sdd), rate = prior_rate) where internal_rate is given a N(0,1) prior."

Just a quick question: would a normal prior be okay for log_sdd? Or a log-normal prior on sdd (and fitting sdd rather than exp(log_sdd))?

Cheers, Richard.

**Comment by finnlindgren** [comment]
First, about the priors, every component defines a latent Gaussian variable (or set of variables), and inlabru iteratively linearises the predictor with respect to those variables. "hyper"parameters of those Gaussians control (co)variance properties, and those parameters can have different priors.  So components that are used to give non-linear behaviour are usually restricted to having Gaussian priors, as they are defined as latent Gaussian components. But due to the nonlinear expressions, you can transform a gaussian into a different distribution. A component `z~N(m,s^2)` will "act" as having a log-normal distribution if it's used as `exp(z)` in the predictor expression. The issue with using log-normal priors for some parameters is that it's a relatively heavy tailed distribution, so other transformations may be preferable. In general, you can use any R quantile function to convert a standardised N(0,1) variable into any other continuous distribution, like the Exponential example I gave in the email you quoted above.  But you cannot set a non-negative distribution (like log-normal) prior on the std.dev. parameter directly, since it needs to be defined as a latent Gaussian variable to work with the iterative linearisation method; you need to explicitly transform it.

There are some helper functions for defining numerically safer methods than just doing `qexp(pnorm(z, ...))` that tries to use as exact methods as possible in the tails. One approach is to define a separate wrapper function for each parameter prior:
```
sigma_prior_transf <- function(x) {
  # Exponentail distribution with rate parameter prior_rate_value (replace with actual value)
  inlabru::bru_forward_transformation(qexp, x, rate = prior_rate_value)
}

comp <- ~ -1 + h(1) + meand(1) + internal_sigma(1, prec.linear = 1)
formula <- normcurve ~ h * exp(-((x - meand)^2 / (2 * sigma_transf(internal_sigma)^2))
```
[Edit: having an exponential prior on sigma makes sense when it's an actual standard deviation parameter. In the use here, it really isn't that, so some other prior may be more appropriate, potentially including a log-normal prior! I would still recommend implementing that with a transformation wrapper instead of a plain `exp(...)` in the formula, to make it easier to change it if/when needed.]

For the Binomial observation model, you have to take into account the INLA link function for `family="binomial"` as well. Ideally, you would want to be able to set an identity link function. This is possible, but unfortunately the inlabru linearisation will not guarantee a binomial probability between 0 and 1, so this won't work in practice.  Instead, one can try to "cancel out" the logit link in the predictor. Note that `qlogis` is the logit transformation and `plogis` is its inverse. Then we can do this:
```
formula <- response ~ qlogis( h * exp(-((x - meand)^2 / (2 * sigma_transf(internal_sigma)^2)) )
```
which is a basic way of giving
```
eta = our "linear" predictor
prob = plogis(eta) = plogis(qlogis("the normal curve")) = "the normal curve"
```
so that the combined effect is that it tries to use the "normal curve" as probability in the Binomial observation model.
However, we can see that this only works if `h` is constrained to be between 0 and 1, so you will need to make a transformation for that parameter as well to ensure the model is properly defined.

For specific cases, one can most likely construct more numerically stable versions of such predictor expressions; e.g. by using that
```
logit(p) = log(p) - log(1-p)
         = log(h*exp(something)) - log(1-h*exp(something)) 
         = log(h) + something - log(1-h*exp(something)) 
```
which is more numerically accurate/stable in some cases (and one can use the `log1p()` function to do `log1p(-h*exp(something))` which may also help). [Edit: the thing to look out for in assessing whether there is likely to be numerical issues is when there is risk of numerical overflow or underflow, e.g. in `exp()` transformation, and to avoid them when/if possible; in the example here only one `exp(something)` is avoided, so it's probably not worth the extra complexity, compared with just using `qlogis`.]
